drivers/usb/device/usb_dc_stm.c

```diff
@@ -0,0 +1,620 @@
+/* USB device controller driver for STM32 devices */
+
+/*
+ * Copyright (c) 2017 Christer Weinigel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @file
+ * @brief USB device controller driver for STM32 devices
+ *
+ * This driver uses the STM Cube low level drivers to talk to the USB
+ * device controller on the STM32 family of devices using the
+ * STM32Cube HAL layer.
+ *
+ * There is a bit of an impedance mismatch between the Zephyr
+ * usb_device and the STM32 Cube HAL layer where higher levels make
+ * assumptions about the low level drivers that don't quite match how
+ * the low level drivers actuall work.
+ *
+ * The usb_dc_ep_read function expects to get the data it wants
+ * immediately while the HAL_PCD_EP_Receive function only starts a
+ * read transaction and the data won't be available until call to
+ * HAL_PCD_DataOutStageCallback. To work around this I've
+ * had to add an extra packet buffer in the driver which wastes memory
+ * and also leads to an extra copy of all received data.  It would be
+ * better if higher drivers could call start_read and get_read_count
+ * in this driver directly.
+ *
+ * Note, the STM32F4xx series of MCUs seem to use a Synopsys
+ * DesignWare controller which is very similar to the one supported by
+ * usb_dc_dw.c so it might be possible get USB support if that driver
+ * is modified a bit.  On the other hand, using the STM32 Cube drivers
+ * means that this driver ought to work on STM3xxx and maybe also on
+ * some Renesas processors which use a similar HAL.
+ *
+ */
+
+#include <soc.h>
+#include <string.h>
+#include <usb/usb_dc.h>
+#include <usb/usb_device.h>
+#include <clock_control/stm32_clock_control.h>
+
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_USB_DRIVER_LEVEL
+#include <logging/sys_log.h>
+
+/* Allocate FIFO memory evenly between the RX/TX FIFOs (1 RX FIFO and n TX). */
+#define FIFO_EP_SZ_WORD ((CONFIG_USB_DC_STM_FIFO_WORD_NUM) / \
+						(1 + CONFIG_USB_DC_STM_EP_NUM))
```

Comment:
Could we define a FIFO_NUM instead of use (1 + CONFIG_USB_DC_STM_EP_NUM)?

Reply:
Yes but do you agree with the definition. Since it is hardware specific it should be abstracted by the cube HAL at some point in the future.

Reply:
Yes, it should be abstracted. I haven't found any define in stm32cube about number of FiFos or the number of endpoints. We could configure the number of endpoints in Kconfig.defconfig.stm32fxxxxx files of the SOC for now and later in the DT files.